src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AdminControllerTests.cs

```diff
@@ -0,0 +1,218 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using DevAdventCalendarCompetition.Controllers;
+using DevAdventCalendarCompetition.Models.Test;
+using DevAdventCalendarCompetition.Repository.Models;
+using DevAdventCalendarCompetition.Resources;
+using DevAdventCalendarCompetition.Services.Interfaces;
+using DevAdventCalendarCompetition.Services.Models;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Moq;
+using Xunit;
+using static DevAdventCalendarCompetition.Tests.TestHelper;
+
+namespace DevAdventCalendarCompetition.Tests.UnitTests.Controllers
+{
+    public class AdminControllerTests
+    {
+        private readonly Mock<ITestService> _testServiceMock;
+        private readonly Mock<IAdminService> _adminServiceMock;
+
+        public AdminControllerTests()
+        {
+           this._testServiceMock = new Mock<ITestService>();
+           this._adminServiceMock = new Mock<IAdminService>();
+        }
+
+        [Fact]
+        public void Index_ReturnsAViewResultWithAListOfTestDto()
+        {
+            // Arrange
+            var testList = new List<TestDto>()
+            {
+                GetTestDto()
+            };
+
+            this._adminServiceMock.Setup(x => x.GetAllTests()).Returns(testList);
+
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+
+            // Act
+            var result = controller.Index();
+
+            // Assert
+            var viewResult = Assert.IsType<ViewResult>(result);
+            var model = Assert.IsAssignableFrom<List<TestDto>>(viewResult.ViewData.Model);
+            Assert.Single(model);
+        }
+
+        [Fact]
+        public void AddTest_ModelStateIsInvalid_ReturnsViewWithInvalidModel()
+        {
+            // Arrange
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+            controller.ModelState.AddModelError("Number", "Required");
+
+            // Act
+            var result = controller.AddTest(GetTestViewModel());
+
+            // Assert
+            var viewResult = Assert.IsType<ViewResult>(result);
+            Assert.IsType<TestViewModel>(viewResult.ViewData.Model);
```

Comment:
we should check more - the model error
```
Assert.IsTrue(! controller.ModelState.IsValid);
Assert.IsTrue(  controller.ModelState["Number"].Errors.Any( modelError => modelError.ErrorMessage == "Required"));
```

Reply:
If we have an invalid model, we're just returning view with the model, and nothing than that indicates that we have an error, so we only check if we returning that view. We can assert model state error, but with unit test we'll be testing line 56. What do u think?